schema/encoding/jsonschema.go

```diff
@@ -0,0 +1,149 @@
+package jsonschema
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	"github.com/rs/rest-layer/schema"
+)
+
+type errWriter struct {
+	w   io.Writer
+	err error
+}
+
+func (ew errWriter) writeFormat(format string, a ...interface{}) {
+	if ew.err != nil {
+		return
+	}
+	_, ew.err = fmt.Fprintf(ew.w, format, a...)
+}
+
+func (ew errWriter) writeString(s string) {
+	if ew.err != nil {
+		return
+	}
+	_, ew.err = ew.w.Write([]byte(s))
+}
+
+func (ew errWriter) write(b []byte) {
+	if ew.err != nil {
+		return
+	}
+	_, ew.err = ew.w.Write(b)
+}
+
+// ValidatorToJSONSchema takes a validator and renders to JSON
+func validatorToJSONSchema(w io.Writer, v schema.FieldValidator) (err error) {
+	if v == nil {
+		return nil
+	}
+	ew := errWriter{w: w}
+	switch t := v.(type) {
+	case *schema.String:
+		ew.writeString(`"type": "string"`)
+		if t.Regexp != "" {
+			ew.writeFormat(`, "pattern": %q`, t.Regexp)
+		}
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, fmt.Sprintf("%q", value))
+			}
+			ew.writeFormat(`, "enum": [%s]`, strings.Join(allowed, ", "))
+		}
+		if t.MinLen > 0 {
+			ew.writeFormat(`, "minLength": %s`, strconv.FormatInt(int64(t.MinLen), 10))
+		}
+		if t.MaxLen > 0 {
+			ew.writeFormat(`, "maxLength": %s`, strconv.FormatInt(int64(t.MaxLen), 10))
+		}
+	case *schema.Integer:
+		ew.writeString(`"type": "number"`)
+
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, strconv.FormatInt(int64(value), 10))
+			}
+			ew.writeFormat(`, "enum": [%s]`, strings.Join(allowed, ","))
+		}
+		if t.Boundaries != nil {
+			ew.writeFormat(`, "minimum": %s, "maximum": %s`,
+				strconv.FormatFloat(t.Boundaries.Min, 'E', -1, 64),
+				strconv.FormatFloat(t.Boundaries.Max, 'E', -1, 64))
+		}
+	case *schema.Array:
+		ew.writeString(`"type": "array"`)
+		if t.ValuesValidator != nil {
+			ew.writeString(`, "items": `)
+			if ew.err == nil {
+				ew.err = validatorToJSONSchema(w, t.ValuesValidator)
+			}
+		}
+	case *schema.Object:
+		if ew.err == nil {
+			ew.err = schemaToJSONSchema(w, t.Schema)
+		}
+	case *schema.Time:
+		ew.writeString(`"type": "string", "format": "date-time"`)
```

Comment:
Let's get this first tranche merged and then you can submit specific PRs for things you are interested in. Sound good? I just added Float since it was easy to do, and it looks very similar to Integer.